crates/block-producer/src/mempool/mod.rs

```diff
@@ -336,14 +349,64 @@ impl Mempool {
     ///
     /// Panics if there is already a block in flight.
     #[instrument(target = COMPONENT, name = "mempool.select_block", skip_all)]
-    pub fn select_block(&mut self) -> (BlockNumber, Vec<ProvenBatch>) {
-        assert!(self.block_in_progress.is_none(), "Cannot have two blocks inflight.");
+    pub fn select_block(&mut self) -> (BlockNumber, Vec<Arc<ProvenBatch>>) {
+        // The selection algorithm is fairly neanderthal in nature.
+        //
+        // We iterate over all proven batch nodes, each time selecting the first which has no
+        // parent nodes that are unselected batches.
+        //
+        // Note that selecting a batch can unblock other batches. This implementation handles this
+        // by resetting the iteration whenever a batch is selected.
 
-        let batches = self.batches.select_block(self.block_budget);
-        self.block_in_progress = Some(batches.iter().map(ProvenBatch::id).collect());
-        self.inject_telemetry();
+        assert!(
+            self.nodes.proposed_block.is_none(),
+            "block {} is already in progress",
+            self.nodes.proposed_block.as_ref().unwrap().0
+        );
+
+        let mut selected = BlockNode::default();
+        let mut budget = self.config.block_budget;
+        let mut candidates = self.nodes.proven_batches.values();
+
+        'next: while let Some(candidate) = candidates.next() {
+            if selected.contains(candidate.id()) {
+                continue 'next;
+            }
+
+            // A batch is selectable if all parents are already blocks, or if the batch is part of
+            // the current block selection.
+            for parent in self.state.parents(NodeId::ProvenBatch(candidate.id()), candidate) {
+                match parent {
+                    NodeId::Block(_) => {},
+                    NodeId::ProvenBatch(parent) if !selected.contains(parent) => {},
+                    _ => continue 'next,
+                }
+            }
```

Comment:
Is this right? We want to move forward with the batch whose parent batches are already within the in-progress block.

```suggestion
            for parent in self.state.parents(NodeId::ProvenBatch(candidate.id()), candidate) {
                match parent {
                    NodeId::Block(_) => {},
                    NodeId::ProvenBatch(parent) if selected.contains(parent) => {},
                    _ => continue 'next,
                }
            }
```

Reply:
Ah yes; thank you! I think I had the logic inverted previously and I missed this during that refactor.